RFC for integrating pcap-release with BOSH (#640)
cloudfoundry/community: toc/rfc/rfc-draft-pcap-bosh.md

# Meta
[meta]: #meta
- Name: Integrate pcap feature into BOSH
- Start Date: 2023-07-06

> - Author(s): @domdom82 @maxmoehl

Reviewer: ```suggestion
- Author(s): [@domdom82](https://github.com/domdom82)  [@maxmoehl](https://github.com/maxmoehl)
```

Author: I don't think those are supposed to be links. Userhandles should be fine. (double check other RFCs, e.g. [here](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0010-stemcell-cleanup.md) )

- Status: Draft
- RFC Pull Request: https://github.com/cloudfoundry/community/pull/640

## Summary

Members of the [App Runtime Platform WG](https://github.com/cloudfoundry/community/blob/main/toc/working-groups/app-runtime-platform.md) are developing a BOSH release called [pcap-release](https://github.com/cloudfoundry/pcap-release) which adds network traffic capturing tools to both BOSH deployments and Cloud Foundry applications.
The release was created after an initial discussion was started on the [cf-deployment](https://github.com/cloudfoundry/cf-deployment/issues/980) repository.

> Since then, the project has evolved into a BOSH use case and a Cloud Foundry use case. This RFC discusses a proposal to get the BOSH use case integrated with BOSH director and BOSH cli.

Reviewer: ```suggestion
Since then, the project has evolved into a BOSH use case and a Cloud Foundry use case. This RFC discusses a proposal to get the BOSH use case integrated with BOSH director, BOSH agent and BOSH cli.
```

The main goal of the pcap feature is to allow users (BOSH operator in this case) to easily capture network traffic across multiple VMs by specifying various parameters in PCAP CLI:
- BOSH deployment
- BOSH instance group(s)
- BPF [Berkeley Packet Filter](https://www.tcpdump.org/manpages/pcap-filter.7.html) filter to apply

PCAP CLI will form a capture request and send it to PCAP API. The capture will then take place in real-time similar to running [tcpdump](https://www.tcpdump.org/manpages/tcpdump.1.html) in a distributed fashion across multiple machines. The PCAP API will contact PCAP AGENT on all nodes involved and trigger a network capture. The PCAP AGENTs will concurrently stream all captured traffic back to PCAP AGENT where all streams are merged into one contiguous stream that is sent back to the PCAP CLI where it is then stored as a PCAP FILE on disk.

> 

Reviewer: I think it's important to mention that we _have to_ use HTTP/2 listeners. Consequentially, the integration would need to support those as well.
```suggestion

All communication between PCAP CLI, PCAP API and PCAP AGENT(s) is carried out via gRPC over HTTP/2.

```

Trust is established as follows:
1. PCAP CLI uses the BOSH CLI `access_token` and sends them to PCAP API

> 2. PCAP API verifies `access_token` against BOSH UAA and establishes mTLS with PCAP AGENT

Reviewer: ```suggestion
2. PCAP API verifies `access_token` against BOSH UAA, verifies the token scope and establishes mTLS with PCAP AGENT
```

3. PCAP AGENT only accepts mTLS connections from PCAP AGENT via a trusted CA

## Problem

The current architecture for "PCAP for BOSH" is depicted as follows:

![PCAP BOSH Architecture](rfc-draft-pcap-bosh/tcpdump-for-bosh.svg)

There are several flaws with this architecture that are the result of missing integration:
1. The PCAP BOSH CLI does not share context with BOSH CLI (because unlike CF CLI, BOSH CLI has no plugin mechanism).

> 2. PCAP BOSH CLI reads the user's `access_token` and `refresh_token`, talk to BOSH UAA to get a new `access_token` using the `refresh_token` and write both back to `.bosh/config` - a file which is also accessed by BOSH CLI.

Reviewer: ```suggestion
2. PCAP BOSH CLI reads the user's `access_token` and `refresh_token`, talks to BOSH UAA to get a new `access_token` using the `refresh_token` and writes both back to `.bosh/config` - a file which is also accessed by BOSH CLI.
```

3. The `access_token` is then sent to PCAP API which uses it as a credential to talk to BOSH DIRECTOR
4. PCAP API needs to speak to BOSH DIRECTOR in order to find the endpoints (IP:PORT) of each PCAP AGENT needed for the capture.
5. In order to make this feature work, PCAP AGENT needs to be available on all BOSH managed VMs. The only way today would be to use it as an [addon](https://bosh.io/docs/runtime-config/#addons)

## Proposal

### Integrate PCAP CLI as a "pcap" keyword into BOSH CLI

- This would allow PCAP BOSH CLI to share context with BOSH CLI and have access to the user's login session as well as the currently used BOSH environment.

- Another benefit would be that users would not have to download the PCAP BOSH CLI separately. 

### Co-locate PCAP AGENT with BOSH AGENT

> - This would make network capture a "built-in" feature of BOSH. It removes the need to manually add PCAP AGENT as a job to deployment manifests or configure it as an addon.

Reviewer 2: On agent side the proposal is to have the `pcap` functionality integrated directly into the BOSH agent by importing our library. Unless there is a compelling reason not to do so?

Author: Fair point, I didn't know they rewrote it from ruby to go already. I will rephrase to "integrate with BOSH AGENT".
The details of the integration (submodule or not etc.) will be discussed with the community. We only need to maintain the ability to also have pcap-agent source on a separate repo so that it may be integrated with CF also.

Who knows, could also become a separate "pcap-bosh-agent" if integrating while retaining original sources is not a feasible option.

Reviewer 2: We will not split the source. All go code related to pcap will remain in a single repository (and go module), in this case we would probably just have a dependency to that module from the bosh-agent.

> ### Co-locate PCAP API with BOSH DIRECTOR

Author: I'd say we should mention this as a hook for additional discussions with the community. That's why I would like to include the option in the RFC so that it won't be dismissed later. The goal of the RFC should be to trigger a discussion and address possible options and concerns, not provide a finite solution. 
I think the "director bottleneck" caveat will be definitively discussed. If you can think of more aspects that could need addressing, feel free to add them as you see fit.

> ### Co-locate PCAP API with BOSH DIRECTOR

Reviewer 3: Added a proposal to include the "singleton" aspect to the existing caveat

> - This would make it very easy for PCAP CLI to find the right endpoint of PCAP API (it would be equal to the current `BOSH_ENVIRONMENT`)

Reviewer 2: `BOSH_ENVIRONMENT` is just the alias to look for in the BOSH config. The availability of PCAP API would probably be announced via the `/info` endpoint which is already used to point to services adjacent to the BOSH director.

Author: Agreed. Changed draft accordingly.

> - No route-registrar or gorouter would be needed to announce any routes. The endpoint would be implicitly known via the BOSH CLI.

Reviewer 2: It is not strictly required right now either, you could directly access the API VM. But yes, gorouter would then be completely out of the game.

Author: true. I'd leave it in though to describe the difference of how we would likely deploy today (in conjunction with CF).

> - Possible caveat: Large network dumps could strangle network bandwidth on BOSH DIRECTOR. Therefore an option to have several PCAP API nodes (load-balanced via BOSH DNS) deployed separately may be considered as an alternative.

Reviewer 2: ```suggestion
- Large network dumps could exhaust resources on the BOSH director which wouldn't be a concern with a dedicated, horizontally scalable deployment.
```

Author: I disagree with this suggestion. The goal is to co-locate with director, which would be the most convenient. There is a downside to that convenience and there should be an option to remove that downside. Your suggestion, as it reads, implies that there will be a dedicated deployment to avoid resource exhaustion. That contradicts the overall proposal of "co-locating with BOSH director". My original sentence wanted to retain both options with the co-location as the default.

Author: @domdom82 what about
```
Large network dumps could exhaust resources on the BOSH director, therefore an option for a dedicated, horizontally scalable deployment should exist.
```
?

> - Large network dumps could exhaust resources on the BOSH director, therefore an option for a dedicated, horizontally scalable deployment should exist.

Reviewer 3: ```suggestion
- Large network dumps could exhaust resources on the BOSH director and this approach adds a runtime dependency to the singleton BOSH director, therefore an option for a dedicated, horizontally scalable deployment should exist.
```
